relay-metrics/src/aggregatorservice.rs

```diff
@@ -128,144 +94,250 @@ pub struct FlushBuckets {
-                self.process_batches(buckets, |batch| {
-                    if let Some(ref receiver) = self.receiver {
-                        receiver.send(FlushBuckets {
-                            project_key,
-                            partition_key,
-                            buckets: batch,
-                        });
-                    }
-                });
+                let capped_batches =
+                    CappedBucketIter::new(buckets.into_iter(), self.max_flush_bytes);
+
+                let num_batches = capped_batches
+                    .map(|batch| {
+                        relay_statsd::metric!(
+                            histogram(MetricHistograms::BucketsPerBatch) = batch.len() as f64,
+                            aggregator = aggregator_name,
+                        );
+                        if let Some(receiver) = self.receiver.as_ref() {
+                            receiver.send(FlushBuckets {
+                                project_key,
+                                partition_key,
+                                buckets: batch,
+                            });
+                        };
+                    })
+                    .count();
+
+                relay_statsd::metric!(
+                    histogram(MetricHistograms::BatchesPerPartition) = num_batches as f64,
+                    aggregator = aggregator_name,
+                );
             }
         }
         relay_statsd::metric!(
             histogram(MetricHistograms::BucketsFlushed) = total_bucket_count,
-            aggregator = self.aggregator.name(),
+            aggregator = aggregator_name,
         );
     }
 
-    fn handle_merge_buckets(&mut self, msg: MergeBuckets) {
-        let MergeBuckets {
-            project_key,
-            buckets,
-        } = msg;
-        self.aggregator
-            .merge_all(project_key, buckets, self.max_total_bucket_bytes);
+    /// Flushes all the aggregators and sends the [`FlushBuckets`] message to the receiver
+    /// in the fire and forget fashion. It is up to the receiver to send the [`MergeBuckets`]
+    /// message back if buckets could not be flushed and we require another re-try.
+    ///
+    /// If `force` is true, flush all buckets unconditionally and do not attempt to merge back.
+    fn try_flush_all(&mut self) {
+        let force_flush = matches!(&self.state, &AggregatorState::ShuttingDown);
+
+        let flush_buckets: Vec<_> = self
+            .aggregators_mut()
+            .map(|agg| agg.pop_flush_buckets(force_flush))
+            .collect_vec();
+
+        // Was unable to do it in one line due to borrow checker constraints.
+        for (name, buckets) in self
+            .aggregators_ref()
+            .map(|agg| agg.name())
+            .zip(flush_buckets.into_iter())
+        {
+            if buckets.is_empty() {
+                continue;
+            };
+
+            self.try_flush(name, buckets);
+        }
```

Comment:
We need mutable access to the aggregators to take out the flush buckets, so I couldn't take out the flush buckets and the aggregator names (&str) at the same time, so this is the best i came up with it. If anyone know a better way then feel free to give it a go.  
 
It sucks we need to take out the names in the first place, it's only used for metrics and has no functional purpose